xla/service/gpu/fusions/transpose_mlir.cc

```diff
@@ -113,14 +167,35 @@ MlirTransposeFusion::MlirTransposeFusion(const HloFusionAnalysis& analysis)
       transposes_to_tile.insert(hero);
       shmem_transpose_roots_.push_back(root);
       shmem_transpose_root_indices_.push_back(index);
-      permutation_ = transpose->permutation;
     } else {
       side_output_roots_.push_back(root);
       side_output_root_indices_.push_back(index);
     }
     ++index;
   }
   shmem_transposes_ = {transposes_to_tile.begin(), transposes_to_tile.end()};
+  const auto& tiled_transpose = analysis.tiled_transpose();
+  absl::c_copy(tiled_transpose.permutation, std::back_inserter(permutation_));
+  permutation_.insert(absl::c_find(permutation_, kPreTransposeMinorDim) + 1,
+                      kPreTransposeVectorizedDim);
+}
+
+// Return bitcast mapping bewteen block_tile and permuted_block_tile if
+// `is_reverse` is false Otherwise return the reverse mapping.
+IndexingMap GetPermuteBlockTileBitcastMap(
+    const Tiling& tiling, const absl::InlinedVector<int64_t, 4>& permutation,
+    MLIRContext* mlir_context, bool is_reverse = false) {
```

Comment:
It's not very clear what "reverse" means. Consider giving this a more descriptive name, or creating two functions instead of one.